inmoose/edgepy/DGEExact.py

```diff
@@ -16,9 +16,28 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 # -----------------------------------------------------------------------------
 
+import pandas as pd
 
-class DGEExact:
-    def __init__(self, table, comparison, genes):
-        self.table = table
+from ..diffexp import DEResults
+
+
+class DGEExact(DEResults):
+    _metadata = ["comparison", "genes"]
+
+    @property
+    def _constructor(self):
+        def f(*args, **kwargs):
+            return DGEExact(
+                *args, comparison=self.comparison, genes=self.genes, **kwargs
+            )
+
+        return f
```

Comment:
Why is it necessary to specify that here?

Reply:
Basically, `DGEExact` is a `pandas.DataFrame` with 2 extra attributes (`comparison` and `genes`).
Its constructor (`__init__`) takes any arguments valid for `pandas.DataFrame.__init__` and forwards them to `pandas.DataFrame.__init__`, plus 2 arguments to initialize the 2 extra attributes.

The `_constructor` property returns the **constructor** to use when building a sub-frame _e.g._ through indexing, `.loc` or `.iloc`. The returned constructor is automatically passed the arguments used to build the sub-`pandas.DataFrame`, but here we want to pass it two extra arguments (for the 2 extra attributes).
This is what `f` is for: passing `self.comparison` and `self.genes` on to the sub-`DGEExact` being built.

For all the nitty-gritty details, have a look at [pandas documentation](https://pandas.pydata.org/docs/development/extending.html#subclassing-pandas-data-structures).

Reply:
Ok thanks!